### PR TITLE
Task/migrate esper 8.4

### PR DESCRIPTION
**Pull request by AlvaroVega**
issue #136 

- [x] Clean old code
- [x]  Improve logs
- [x] Add new tests: https://github.com/telefonicaid/perseo-core/blob/8e2420e123a9c516b2f52c75c92797dd13f37277/perseo-main/src/test/java/com/telefonica/iot/perseo/test/Help.java#L53-L56
- [x] Select minor version of esper (> 8.4 ```com.espertech.esper.compiler.client.EPCompileExceptionItem: Implicit conversion from datatype 'double' to 'Float' for property 'cast(cast(BloodPressure?,String),float)' is not allowed (strict filter type coercion)```
- [x] Propagate error detail to perseo-fe

### CODE REVIEW

**Comment by AlvaroVega** [comment]
sudo docker exec iotp-pqa1 bash -c "cd /opt/iotp-pqa && behave ./tests/ --tags=onpremise --tags=-skip --tags=cb2cep"

Failing scenarios:
  tests/features/TS03_cbflows/001_cb2cep.feature:91  cb001003 Entity update in CB matches a basic TYPE rule with action EMAIL -- @1.1 
  tests/features/TS03_cbflows/001_cb2cep.feature:92  cb001003 Entity update in CB matches a basic TYPE rule with action EMAIL -- @1.2 
  tests/features/TS03_cbflows/001_cb2cep.feature:261  cb001010 Entity update in CB matches a basic TYPE rule with action EMAIL and wildcards -- @1.1 
  tests/features/TS03_cbflows/001_cb2cep.feature:262  cb001010 Entity update in CB matches a basic TYPE rule with action EMAIL and wildcards -- @1.2 
  tests/features/TS03_cbflows/006_cb2cep2_conditionbasedonattrmetadata.feature:109  trigger a plain rule with a Email action -- @1.1 
  tests/features/TS03_cbflows/006_cb2cep2_conditionbasedonattrmetadata.feature:110  trigger a plain rule with a Email action -- @1.2 
  tests/features/TS03_cbflows/006_cb2cep2_conditionbasedonattrmetadata.feature:144  trigger a plain rule with a Email action and Metadata available -- @1.1 
  tests/features/TS03_cbflows/006_cb2cep2_conditionbasedonattrmetadata.feature:145  trigger a plain rule with a Email action and Metadata available -- @1.2 
  tests/features/TS03_cbflows/006_cb2cep2_conditionbasedonattrmetadata.feature:251  trigger a plain rule with a post action -- @1.2 
  tests/features/TS03_cbflows/008_cb2cep2_conditionbasedonlocation.feature:79  trigger a plain rule based on location using GeoJSON format and Email action
  tests/features/TS03_cbflows/009_cb2cep2_conditionbasedontimestamp.feature:72  SC_2 - cb00902 - Trigger a plain rule based on timestamp if it is into the 30 last minutes and Email action

**Comment by AlvaroVega** [comment]
with esper 8.5+ EPL like
```
 "@Audit select *,\"blood_1_action\" as iotcepaction,"
                    + "ev.BloodPressure? as Pression, ev.id? as Meter from pattern "
                    + "[every ev=iotEvent(cast(BloodPressure?,String)>1.5"
                    + " and type=\"BloodMeter\")]"
```
com.espertech.esper.compiler.client.EPCompileExceptionItem: Implicit conversion from datatype 'double' to 'Float' for property 'cast(cast(BloodPressure?,String),float)' is not allowed (strict filter type coercion)

**Comment by AlvaroVega** [comment]
Enabling query plan:
        // Produce query plan output unless logging at debug-level
        configuration.getCommon().getLogging().setEnableQueryPlan(true);

and trying with a rule with text: 

expression double js:fib(num) [ fib(num); function fib(n) { if(n <= 1) return n; return fib(n-1) + fib(n-2);} ] expression twoPI alias for { java.lang.Math.PI * 2 } SELECT "mirule2" as ruleName, *, 3 as r FROM iotEvent 


time=2021-09-16T15:05:51.852Z | lvl=DEBUG | from=::ffff:172.17.0.15 | corr=dabb741a-a074-4fdd-84eb-79f783c0da01; node=k6ah1LRew; perseocep=67 | trans=a3f9e91e-8fff-4419-bed7-806fe534b587 | srv=smartcity | subsrv=/ | op=parse | comp=perseo-core | msg=Error parsing statement [expression double js:fib(num) [ fib(num); function fib(n) {  if(n <= 1) return n; return fib(n-1) + fib(n-2);} ] expression twoPI alias for { java.lang.Math.PI * 2 }  context ctxt$smartcity$ SELECT "mirule2" as ruleName, *, 3 as r FROM iotEvent]
org.antlr.v4.runtime.misc.ParseCancellationException
	at org.antlr.v4.runtime.BailErrorStrategy.recover(BailErrorStrategy.java:51)
	at com.espertech.esper.compiler.internal.generated.EsperEPL2GrammarParser.expressionDef(EsperEPL2GrammarParser.java:1054)
	at com.espertech.esper.compiler.internal.generated.EsperEPL2GrammarParser.expressionDecl(EsperEPL2GrammarParser.java:925)
	at com.espertech.esper.compiler.internal.generated.EsperEPL2GrammarParser.startEPLExpressionRule(EsperEPL2GrammarParser.java:641)
	at com.espertech.esper.compiler.internal.util.CompilerHelperSingleEPL$1.invokeParseRule(CompilerHelperSingleEPL.java:49)
	at com.espertech.esper.compiler.internal.parse.ParseHelper.parse(ParseHelper.java:94)
	at com.espertech.esper.compiler.internal.util.CompilerHelperSingleEPL.parse(CompilerHelperSingleEPL.java:155)
	at com.espertech.esper.compiler.internal.util.CompilerHelperSingleEPL.parseCompileInlinedClassesWalk(CompilerHelperSingleEPL.java:62)
	at com.espertech.esper.compiler.internal.util.CompilerHelperStatementProvider.compileItem(CompilerHelperStatementProvider.java:92)
	at com.espertech.esper.compiler.internal.util.CompilerHelperModuleProvider.compileToBytes(CompilerHelperModuleProvider.java:127)
	at com.espertech.esper.compiler.internal.util.CompilerHelperModuleProvider.compile(CompilerHelperModuleProvider.java:89)
	at com.espertech.esper.compiler.internal.util.EPCompilerImpl.compile(EPCompilerImpl.java:98)
	at com.telefonica.iot.perseo.Utils.compileDeploy(Utils.java:417)
	at com.telefonica.iot.perseo.RulesManager.make(RulesManager.java:138)
	at com.telefonica.iot.perseo.RulesServlet.doPost(RulesServlet.java:122)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:681)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:764)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:196)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:544)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:698)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:364)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:624)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:831)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1650)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.antlr.v4.runtime.NoViableAltException
	at com.espertech.esper.compiler.internal.generated.EsperEPL2GrammarParser.stringconstant(EsperEPL2GrammarParser.java:22593)
	at com.espertech.esper.compiler.internal.generated.EsperEPL2GrammarParser.expressionDef(EsperEPL2GrammarParser.java:1042)
	... 37 more


only sencond expression works

File: perseo-main/src/test/java/com/telefonica/iot/perseo/test/Help.java [hunk 1]
```diff
@@ -44,29 +44,54 @@
...
+                 ,"SELECT current_timestamp() as r FROM iotEvent WHERE cast(value?,int) > 10"
+                 ,"expression twoPI alias for { java.lang.Math.PI * 2 > 5 } SELECT *, twoPI as r FROM iotEvent WHERE twoPI"
+                 ,"expression E {(v1,v2) => max(v1,v2)} select E(1, 2) from iotEvent"
+                 ,"expression E {(v) => case when v is null or cast(v,float) < 0 or cast(v,string) = 'null' or cast(v,string) = ' ' then 0 else v end}  select E(5) from iotEvent"
+                 ,"select current_timestamp() as r from iotEvent where (cast(price?,String) != '321' and (cast(price?,String) != '123'))"
+                 //,"inlined_class \"\"\"\n  public class MyUtility {\n    public static double fib(int n) {\n      if (n <= 1) {\n        return n;\n      }\n      return fib(n-1) + fib(n-2);\n    }\n  }\n\"\"\"\n expression FIBC alias for { MyUtility.fib(5) > 3 } SELECT *, FIBC as r FROM iotEvent WHERE FIBC"
+                 //,"expression double js:fib(num) [\nfib(num);\nfunction fib(n) {\n  if(n <= 1)\n    return n;\n  return fib(n-1) + fib(n-2);\n}\n]\nexpression FIBE alias for { fib(5) > 3 } SELECT *, FIBE as r FROM iotEvent WHERE FIBE"
```
Comment (fgalan) [inline perseo-main/src/test/java/com/telefonica/iot/perseo/test/Help.java:59]: Leftover commented lines?

File: perseo-main/src/main/java/com/telefonica/iot/perseo/RulesManager.java [hunk 2]
```diff
@@ -114,33 +127,52 @@ public static synchronized Result make(EPServiceProvider epService, String text)
...
+                                                                                                )));
                     statement.addListener(new GenericListener());
                 } else {
                     logger.debug(String.format("found repeated statement: %s", name));
                     statement = prevStmnt;
                 }
             }
             return new Result(HttpServletResponse.SC_OK,
-                    Utils.Statement2JSONObject(statement).toString());
+                              Utils.Statement2JSONObject(statement, epa, statement.getDeploymentId()).toString());
```
Comment (fgalan) [inline perseo-main/src/main/java/com/telefonica/iot/perseo/RulesManager.java:175]: Over indentation here?

File: perseo-main/src/test/java/com/telefonica/iot/perseo/test/EventBeanMock.java [hunk 3]
```diff
@@ -110,20 +113,36 @@ public EventPropertyGetterIndexed getGetterIndexed(String string) {
...
+        // @Override
+        // public EPTypeClass getUnderlyingEPType() {
+        //     throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
+        // }
+
+        // @Override
+        // public EPType getPropertyEPType(java.lang.String string) {
+        //     throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
+        // }
+        };
```
[the remark this answers is not in the document]
Comment (AlvaroVega) [reply perseo-main/src/test/java/com/telefonica/iot/perseo/test/EventBeanMock.java:145]: Just a comment for the future, when we try to update from 8.4 to 8.7, we will need it

File: perseo-main/src/test/java/com/telefonica/iot/perseo/test/Help.java (hunk 1 above)
Comment (AlvaroVega) [reply perseo-main/src/test/java/com/telefonica/iot/perseo/test/Help.java:59]: Just a comment for the future, we expect that epl text will work with a future version

File: perseo-main/src/main/java/com/telefonica/iot/perseo/RulesManager.java (hunk 2 above)
Comment (AlvaroVega) [reply perseo-main/src/main/java/com/telefonica/iot/perseo/RulesManager.java:175]: fixed in https://github.com/telefonicaid/perseo-core/pull/176/commits/24494fc64018425accd9080f0efee0126e5d7725

File: perseo-main/src/test/java/com/telefonica/iot/perseo/test/EventBeanMock.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (fgalan) [reply perseo-main/src/test/java/com/telefonica/iot/perseo/test/EventBeanMock.java]: NTC

**Review by fgalan — APPROVED** [review]
LGTM

**Pull request merged**